node/overseer/src/lib.rs

```diff
@@ -725,68 +763,159 @@ where
 		let metrics = self.metrics.clone();
 		spawn_metronome_metrics(&mut self, metrics)?;
 
-		// Notify about active leaves on startup before starting the loop
+		let initial_sync_finished = self.sync_oracle.finished_syncing();
+		// Import the active leaves found in the database
 		for (hash, number) in std::mem::take(&mut self.leaves) {
 			let _ = self.active_leaves.insert(hash, number);
 			if let Some((span, status)) = self.on_head_activated(&hash, None).await {
-				let update =
-					ActiveLeavesUpdate::start_work(ActivatedLeaf { hash, number, status, span });
-				self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
+				if initial_sync_finished {
+					// Initial sync is complete. Notify the subsystems and proceed to the main loop
+					let update = ActiveLeavesUpdate::start_work(ActivatedLeaf {
+						hash,
+						number,
+						status,
+						span,
+					});
+					self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
+				}
 			}
 		}
 
-		loop {
+		// If initial sync is not complete - wait for it.
+		// This loop is identical to the main one but doesn't generate `ActiveLeaves` and `BlockFinalized` events until
+		// the initial full sync is complete.
+		// This is an infinite loop which executes only when `!initial_sync_finished`. The weird syntax is only to save
+		// one extra layer if indentation because it's already a bit tough for the eyes.
+		// Think about it like:
+		// ```
+		// if !initial_sync_finished {
+		//		loop {
+		//			select! {
+		//				...
+		//			}
+		// 		}
+		// }
+		//```
```

Comment:
nit: is there a more elegant way of solving this without duplicating code? :)

Reply:
Huh I don't like it either :( I extracted as much logic as possible in functions which are called in both loops.

Maybe I can put the loop in a function and pass `BlockImported` and `BlockFinalized` event handlers as lambdas. Then I'll call the function twice with different event handlers. The rest of the logic is 1:1.

If the overall approach is accepted - I'll give it a shot.

Reply:
You need to write some little state machine. 

```
enum SyncState {
     Scyncing,
     Finished,
}

impl SyncState {
    fn should_forward_event(&mut self, oracle: &SyncOracle) -> bool {
         match self {
             Self::Syncing if oracle.syncing() => false,
             Self::Syncing => { *self = Self::Finished; true },
             Self::Finished => true,
         }
    }
}
````

Then you call `should_forward_event` as part of the `BlockFinalized` and `BlockImported` branches below. This should enable you to stop duplicated all the code.

Reply:
Second loop is no longer required, so this is fixed.